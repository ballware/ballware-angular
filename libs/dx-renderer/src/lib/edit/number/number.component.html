<dx-number-box *ngIf="livecycle.layoutItem && livecycle.layoutItem.options && livecycle.layoutItem.options.dataMember"
  class="mt-2"
  [visible]="(visible.visible$ | async) ?? false"
  [label]="livecycle.layoutItem.options.caption ?? ''"
  [name]="livecycle.layoutItem.options.dataMember"
  [(value)]="value.value"
  [readOnly]="(readonly.readonly$ | async) ?? false"
  [showClearButton]="true"
  [showSpinButtons]="true"
  [min]="options.min"
  [max]="options.max"
  (onFocusIn)="livecycle.onEntered()">
  <dx-validator [validationRules]="(validation.validationRules$ | async) ?? []"></dx-validator>  
</dx-number-box>
