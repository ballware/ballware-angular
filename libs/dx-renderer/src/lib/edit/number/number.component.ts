import { Component, OnInit } from '@angular/core';
import { takeUntil } from 'rxjs';
import { DxNumberBoxModule, DxValidatorModule } from 'devextreme-angular';
import { CommonModule } from '@angular/common';
import { Destroy, EditItemLivecycle, NumberValue, Readonly, Visible } from '@ballware/renderer-commons';
import { Validation, Required } from '../../directives';

export interface NumberItemOptions {  
  min?: number;
  max?: number;
}

@Component({
  selector: 'ballware-edit-number',
  templateUrl: './number.component.html',
  styleUrls: [],
  imports: [CommonModule, DxNumberBoxModule, DxValidatorModule],
  hostDirectives: [Destroy, { directive: EditItemLivecycle, inputs: ['initialLayoutItem'] }, NumberValue, Readonly, Validation, Required, Visible],
  standalone: true
})
export class EditLayoutNumberComponent implements OnInit {

  public options: NumberItemOptions = {};

  constructor(
    public destroy: Destroy,
    public livecycle: EditItemLivecycle,
    public visible: Visible,
    public readonly: Readonly,
    public value: NumberValue,
    public validation: Validation
  ) {}

  ngOnInit(): void {

    this.livecycle.preparedLayoutItem$
      .pipe(takeUntil(this.destroy.destroy$))
      .subscribe((layoutItem) => {        
        this.options = layoutItem?.options?.itemoptions as NumberItemOptions ?? {};

        this.livecycle.registerOption('min', () => this.options?.min, (value) => this.options.min = value as number);
        this.livecycle.registerOption('max', () => this.options?.max, (value) => this.options.max = value as number);
      });  
  }  
}
