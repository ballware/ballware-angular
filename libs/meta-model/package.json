{
  "name": "@ballware/meta-model",
  "version": "8.0.1",
  "type": "commonjs"
}
