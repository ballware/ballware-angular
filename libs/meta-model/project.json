{
  "name": "meta-model",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/meta-model/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/libs/meta-model",
        "main": "libs/meta-model/src/index.ts",
        "tsConfig": "libs/meta-model/tsconfig.lib.json",
        "assets": [
          "libs/meta-model/*.md"
        ]
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs meta-model {args.ver} {args.tag}"
      },
      "dependsOn": [
        "build"
      ]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "libs/meta-model/**/*.ts"
        ]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}"
      ],
      "options": {
        "jestConfig": "libs/meta-model/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
