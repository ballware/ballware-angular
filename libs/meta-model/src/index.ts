export * from './lib/cruditem';
export * from './lib/document';
export * from './lib/entity';
export * from './lib/page';
export * from './lib/queryparams';
export * from './lib/rights';
export * from './lib/scriptutil';
export * from './lib/statistic';
export * from './lib/template';
export * from './lib/tenant';

