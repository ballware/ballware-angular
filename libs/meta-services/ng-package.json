{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/meta-services",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
