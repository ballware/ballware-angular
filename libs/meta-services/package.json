{
  "name": "@ballware/meta-services",
  "version": "8.0.6",
  "peerDependencies": {
    "@angular/common": "^14.2.0",
    "@angular/core": "^14.2.0"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  }
}
