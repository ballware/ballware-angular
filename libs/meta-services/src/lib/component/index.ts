import { StoreModule } from '@ngrx/store';
import { componentReducer } from './component.reducer';
import { componentFeatureKey } from './component.state';

export * from './component.actions';
export * from './component.state';

export const ComponentFeatureModule = StoreModule.forFeature(componentFeatureKey, componentReducer);