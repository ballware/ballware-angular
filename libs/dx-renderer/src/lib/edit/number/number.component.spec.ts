import { ComponentFixture, TestBed } from '@angular/core/testing';

import { EditLayoutNumberComponent, NumberItemOptions } from './number.component';
import { Provider } from '@angular/core';
import { EDIT_SERVICE } from '@ballware/meta-services';
import { EditLayoutItem } from '@ballware/meta-model';
import { mockedEditServiceContext } from '../../../test/editservice.spec';

describe('EditLayoutNumberComponent', () => {
  let component: EditLayoutNumberComponent;
  let fixture: ComponentFixture<EditLayoutNumberComponent>;

  const mockedEditService = mockedEditServiceContext();
        
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ EditLayoutNumberComponent ],
      providers: [        
        {
          provide: EDIT_SERVICE,
          useFactory: () => mockedEditService.mock.object()
        } as Provider         
      ]
    })
    .compileComponents();
  });

  it('should create', () => {
    fixture = TestBed.createComponent(EditLayoutNumberComponent);

    const layoutItem = {
      type: 'text',
      options: {
        dataMember: 'mockedmember',
        itemoptions: {
          min: 0,
          max: 100
        } as NumberItemOptions
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    fixture.componentRef.setInput('initialLayoutItem', layoutItem);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should apply options', () => {
    fixture = TestBed.createComponent(EditLayoutNumberComponent);

    const layoutItem = {
      options: {
          dataMember: 'mockedmember',
          required: false,
          readonly: false,            
          visible: false
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    fixture.componentRef.setInput('initialLayoutItem', layoutItem);
    fixture.detectChanges();

    expect(component.livecycle.getOption('value')).toBe(0);
    expect(component.livecycle.getOption('required')).toBe(false);
    expect(component.livecycle.getOption('readonly')).toBe(false);
    expect(component.livecycle.getOption('visible')).toBe(false);

    component.livecycle.setOption('value', 42);
    expect(component.livecycle.getOption('value')).toBe(42);

    component.livecycle.setOption('required', true);
    expect(component.livecycle.getOption('required')).toBe(true);

    component.livecycle.setOption('readonly', true);
    expect(component.livecycle.getOption('readonly')).toBe(true);

    component.livecycle.setOption('visible', true);
    expect(component.livecycle.getOption('visible')).toBe(true);

    component.livecycle.setOption('min', 2);
    expect(component.livecycle.getOption('min')).toBe(2);

    component.livecycle.setOption('max', 98);
    expect(component.livecycle.getOption('max')).toBe(98);

    expect(() => component.livecycle.getOption('undefined')).toThrowError('Unsupported option <undefined>');
    expect(() => component.livecycle.setOption('undefined', 'any value')).toThrowError('Unsupported option <undefined>');
  });
});
