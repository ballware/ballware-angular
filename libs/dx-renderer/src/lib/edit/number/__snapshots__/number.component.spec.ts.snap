// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLayoutNumberComponent should create 1`] = `
<ballware-edit-number
  destroy={[Function Destroy]}
  livecycle={[Function EditItemLivecycle]}
  options={[Function Object]}
  readonly={[Function Readonly]}
  validation={[Function Validation]}
  value={[Function NumberValue]}
  visible={[Function Visible]}
>
  <dx-number-box
    class="mt-2 dx-validator dx-visibility-change-handler dx-show-invalid-badge dx-numberbox dx-texteditor dx-show-clear-button dx-numberbox-spin dx-editor-outlined dx-widget"
  >
    <input
      name="mockedmember"
      type="hidden"
      value="0"
    />
    <div
      class="dx-texteditor-container"
    >
      <div
        class="dx-texteditor-input-container"
      >
        <input
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="0"
          autocomplete="off"
          class="dx-texteditor-input"
          inputmode="decimal"
          max="100"
          min="0"
          role="spinbutton"
          step="1"
          tabindex="0"
          type="text"
        />
        <div
          class="dx-placeholder dx-state-invisible"
          data-dx_placeholder=""
        />
      </div>
      <div
        class="dx-texteditor-buttons-container"
      >
        <span
          class="dx-clear-button-area"
        >
          <span
            class="dx-icon dx-icon-clear"
          />
        </span>
        <div
          class="dx-numberbox-spin-container"
        >
          <div
            class="dx-widget dx-numberbox-spin-button dx-numberbox-spin-up"
          >
            <div
              class="dx-numberbox-spin-up-icon"
            />
          </div>
          <div
            class="dx-widget dx-numberbox-spin-button dx-numberbox-spin-down"
          >
            <div
              class="dx-numberbox-spin-down-icon"
            />
          </div>
        </div>
      </div>
    </div>
  </dx-number-box>
</ballware-edit-number>
`;
